corporate/views/remote_billing_page.py

```diff
@@ -0,0 +1,175 @@
+import logging
+from typing import Optional, TypedDict
+
+from django.conf import settings
+from django.core import signing
+from django.http import HttpRequest, HttpResponse, HttpResponseRedirect
+from django.shortcuts import render
+from django.urls import reverse
+from django.utils.translation import gettext as _
+from django.views.decorators.csrf import csrf_exempt
+from pydantic import Json
+
+from zerver.lib.exceptions import JsonableError, MissingRemoteRealmError
+from zerver.lib.remote_server import RealmDataForAnalytics
+from zerver.lib.response import json_success
+from zerver.lib.typed_endpoint import PathOnly, typed_endpoint
+from zilencer.models import RemoteRealm, RemoteZulipServer
+
+billing_logger = logging.getLogger("corporate.stripe")
+
+
+SIGNED_ACCESS_TOKEN_VALIDITY_IN_SECONDS = 60 * 60 * 24
+
+
+class RemoteBillingIdentityDict(TypedDict):
+    user_uuid: str
+    user_email: str
+    user_full_name: str
+    remote_server_uuid: str
+    remote_realm_uuid: str
+
+
+@csrf_exempt
+@typed_endpoint
+def remote_server_billing_entry(
+    request: HttpRequest,
+    remote_server: RemoteZulipServer,
+    *,
+    user_uuid: str,
+    user_email: str,
+    user_full_name: str,
+    realm_info: Json[RealmDataForAnalytics],
+) -> HttpResponse:
+    try:
+        remote_realm = RemoteRealm.objects.get(uuid=realm_info.uuid, server=remote_server)
+    except RemoteRealm.DoesNotExist:
+        # This error will prod the remote server to submit its realm info, which
+        # should lead to the creation of this missing RemoteRealm registration.
+        raise MissingRemoteRealmError
+
+    identity_dict = RemoteBillingIdentityDict(
+        user_email=user_email,
+        user_uuid=user_uuid,
+        user_full_name=user_full_name,
+        remote_server_uuid=str(remote_server.uuid),
+        remote_realm_uuid=str(remote_realm.uuid),
+    )
+
+    signed_identity_dict = signing.dumps(identity_dict)
+
+    billing_access_url = settings.ROOT_DOMAIN_URI + reverse(
+        remote_server_billing_finalize_login, args=[signed_identity_dict]
+    )
+    return json_success(request, data={"billing_access_url": billing_access_url})
+
+
+def remote_server_billing_finalize_login(
+    request: HttpRequest, signed_billing_access_token: str
+) -> HttpResponse:
+    try:
+        identity_dict: RemoteBillingIdentityDict = signing.loads(
+            signed_billing_access_token, max_age=SIGNED_ACCESS_TOKEN_VALIDITY_IN_SECONDS
+        )
+    except signing.SignatureExpired:
+        raise JsonableError(_("Billing access token expired."))
+    except signing.BadSignature:
+        raise JsonableError(_("Invalid billing access token."))
+
+    request.session["remote_billing_identity"] = identity_dict
```

Comment:
Is there anything special we need to do with Django sessions to isolate this session variable so that things don't get mixed up with other realms for someone who runs multiple Zulip servers?

One theory would be to make the structure something like `request.session["self_hosting_realms"][remote_realm_uuid] = identity_dict`, so that we'd only look at parts of this session when the URL contains that realm UUID.

We also will need to consider the revocation/expiry story for these sessions.

This detail probably deserves a dedicated chat.zulip.org thread.

Reply:
> One theory would be to make the structure something like request.session["self_hosting_realms"][remote_realm_uuid] = identity_dict, so that we'd only look at parts of this session when the URL contains that realm UUID.

This sounds like a good idea to me - may help us properly support being logged into multiple billing accounts for various servers/realms at once in the future, if we need. Otherwise this will just get overwritten during each auth flow.

Reply:
Does `request.session["remote_billing_identities"][remote_realm_uuid]` make sense? Pushed with that naming for now.

Reply:
Yeah let's do that for this PR; we can revise later.